test/TestConfig/src/org/openj9/envInfo/JavaInfo.java

```diff
@@ -0,0 +1,100 @@
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+package org.openj9.java;
+
+import java.io.*;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+
+public class JavaInfo {
+
+    public String getSPEC() {
+        String osName = System.getProperty("os.name").toLowerCase();
+        String osArch = System.getProperty("os.arch").toLowerCase();
+        String cmprssptrs = System.getProperty("java.fullversion");
+        String spec = "";
+        if (osName.contains("linux")) {
+            spec = "linux";
+        } else if (osName.contains("win")) {
+            spec = "win";
+        } else if (osName.contains("mac")) {
+            spec = "osx";
+        } else if (osName.contains("nix") || osName.contains("nux") || osName.contains("aix")) {
+            spec = "aix";
+        } else if (osName.contains("z/os")) {
+            spec = "zos";
+        } else {
+            spec = osName;
+        }
+
+        if (osArch.contains("amd64") || osArch.contains("x86")) {
+            spec += "_x86";
+        } else if (osArch.contains("ppc") || osArch.contains("powerpc")) {
+            spec += "_ppc";
+        } else if (osArch.contains("s390x")) {
+            spec += "_390";
+        } else if (osArch.contains("aarch")) {
+            spec += "_arm";
+        } else {
+            spec += osArch;
+        }
+
+        if (Integer.parseInt(System.getProperty("sun.arch.data.model").trim())==64) {
+            spec += "-64";
+        }
+
+        if (cmprssptrs != null && cmprssptrs.contains("Compressed References")) {
+            spec += "_cmprssptrs";
+        }
+        if (spec.contains("ppc") && osArch.contains("le")) {
+            spec += "_le";
+        }
+        
+        return spec;
+    }
+
+    public int getJDKVersion() {
+        String javaVersion = System.getProperty("java.version");
+        if (javaVersion.startsWith("1.")) {
+            javaVersion = javaVersion.substring(2);
+        }
+        int dotIndex = javaVersion.indexOf('.');
+        int dashIndex = javaVersion.indexOf('-');
+        return Integer.parseInt(javaVersion.substring(0, dotIndex > -1 ? dotIndex : dashIndex > -1 ? dashIndex : javaVersion.length()));
+    }
+
+
+    public String getJDKImpl() {
+        String impl = System.getProperty("java.vm.vendor").toLowerCase();
+        if (impl.contains("ibm")) {
+            return "ibm";
+        } else if (impl.contains("openj9")) {
+            return "openj9";
+        } else if (impl.contains("oracle") || impl.contains("hotspot") || impl.contains("adoptopenjdk")) {
```

Comment:
Seems that the impl.contains("adoptopenjdk") may not be quite right here, as adopt produces both hotspot and openj9 builds.  Also do not like putting the drop-through 'else' to sap.  We should find out what it returns for java.vm.vendor and check explicitly for it & set it, then the final else we perhaps throw exception "can not determine impl" or set to empty and handle in the callers.

Reply:
For `impl.contains("adoptopenjdk")` part, when I tested on Jenkins using `System.getProperty("java.vm.vendor")`, openj9 gives `openj9`, while hotspot gives `adoptopenjdk`.


For sap part, `System.getProperty("java.vm.vendor")` gives `oracle`, which behaves the same as hotspot. And to handle sap, in `envSettings.mk`, I wrote the following: if the user manually export JDK_IMPL=sap, then the JDKimpl will be set to sap; otherwise the sap impl will be detected and set as hotspot instead.

``` 
ifneq ($(JDK_IMPL), $(DETECTED_JDK_IMPL))
            ifneq ($(JDK_IMPL), sap)
                $(error DETECTED_JDK_IMPL value is $(DETECTED_JDK_IMPL), settled JDK_IMPL value is $(JDK_IMPL). JDK_IMPL value does not match. Please reset or unset JDK_IMPL)
	    endif
endif
```